Implement Ul'xzomit mother and baby distinction

DiscipleOfEris opened the pull request:
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

 * Two thirds of the current Ul'xzomit mobs in Al'Taieu are now babies following the remaining one third Ul'xzomits.
 * The baby Ul'xzomit are smaller and the mother Ul'xzomit are larger.
 * Increase roam distance of mother Ul'xzomit.
 * Mother and baby Ul'xzomit have appropriate (different) drop tables, as in: https://ffxiclopedia.fandom.com/wiki/Ul%27xzomit
    * Make elemental clusters slotted drops (10% for mothers, 1% for babies, with equal chance for each cluster).

### Under the hood

 * Create a new lua function in `scripts/globals/follow.lua` that allows you to call `xi.follow.follow(follower, target[, options])` to have the follower mob persistently follow the target entity.
 * Rename `ROAMFLAG_EVENT` to `ROAMFLAG_SCRIPTED` for a more intuitive name.
 * Add a new `MOBMOD_ROAM_RESET_FACING`. This implemented previous behavior that was hard-coded into `ROAMFLAG_EVENT`. Specifically, it makes mobs reface their original spawn facing when they roamed back home (like after the players died).
 * Add a new `MOBMOD_DONT_ROAM_HOME`. If used, it disables the leash that causes mobs to eventually roam back home.
 * Fix some bugs involving `LookAt()` that cropped up while working on this.
 * Add several utility functions in lua and cpp, and modify the `distance` functions to add an optional `ignoreVertical` flag.
 * Rework the lua `setMobFlags()` function to work from a mob as well, like `mob:setMobFlags(flags)`.
 * Adds a few more `triggerListener()` calls for pathing and roaming.

### Not implemented/Need more research

 * Ul'xzomit roam patterns. At the moment they roam normally but with a larger range. Need more retail research.
 * Babies respawning. At the moment they only spawn when the mother spawns. Need more retail research.
 * Baby roam behavior if mother is dead. At the moment they don't roam at all. Need more retail research.
 * Difference in HP/stats of mother vs. baby. Need more retail research.

Fixes #2604 
Fixes #2573 

~~We might consider making the elemental clusters be slotted drops. At the moment they're implemented as 10% and 1% each for the Ul'xzomit mothers and babies, but I think it's supposed to be 10% and 1% that a cluster drops at all. As is, the mothers have a 57.0% chance for at least one cluster dropping and the babies have a 7.7% chance for at least one cluster.~~

Also, is there interest in a gm command for making a mob follow an entity? I'm not sure if that belongs in LSB core.

## Steps to test these changes

1. Go to Al'Taieu.
2. Watch the Ul'xzomit roam about. The babies follow the mothers.
3. Attack a baby Ul'xzomit. Nothing should link.
4. Attack a mother Ul'xzomit. The two babies should link, but no others.
5. Kill a few mother and baby Ul'xzomit. Notice the different drop rates.

DiscipleOfEris committed:
Add behavior Ul'xzomit mother-baby behaviors

Indicate which Ul'xzomit mobs are mothers with a new LEADER mobMod. Make babies smaller. Make babies link with and follow the mother.

DiscipleOfEris committed:
Add drop tables for Ul'xzomit mothers and babies

Also make the Ul'xzomit mothers larger.

DiscipleOfEris committed:
Add `ROAMFLAG_SCRIPT`

This flag will specify that `mob_controller.cpp` should offload most roam behavior to the entity's `onMobRoamAction()` lua function. It will be called much more frequently than `onMobRoam()` which is called every 3 seconds.

DiscipleOfEris committed:
Fix anomalous worldAngle() results

See #2573

DiscipleOfEris committed:
Refactor follow behavior into separate global function

Create a global function `xi.follow.follow()` so any mob or even gm command can implement the follow behavior. Add several functions to lua `utils` to supplement this and made some minor cpp changes to controllers.

We might consider doing the same for the lua `patrol()` function.

DiscipleOfEris committed:
Fix follow script spewing errors to console

If the leader despawned, the roam script would send errors to console because it didn't exit early like it was supposed to.

Also, clean up a little better when we stop following.

DiscipleOfEris committed:
Make baby Ul'xzomit re-attach

Baby Ul'xzomit will now re-attach when the mother spawns, not just when the babies spawn.

Also clean up duplicate onMobRoamAction() after previous cpp code change.

DiscipleOfEris committed:
Increase Ul'xzomit roam range

Some more work must be done on mother Ul'xzomit to make them more like retail, but I need more information.

TeoTwawki commented on scripts/globals/status.lua:
```diff
@@ -2468,6 +2468,7 @@ xi.mobMod =
     ALLI_HATE           = 68, -- Range around target to add alliance member to enmity list.
     NO_LINK             = 69, -- If set, mob cannot link until unset.
     NO_REST             = 70, -- Mob cannot regain hp (e.g. re-burrowing antlions during ENM).
+    LEADER              = 71, -- Indicate this mob is a leader. Used for Ul'xzomit mobs.
```
`Used for mobs that follow a defined "leader"` would be preferred description.

TeoTwawki commented on src/map/mob_modifier.h:
```diff
@@ -99,6 +99,7 @@ enum MOBMODIFIER : int
     MOBMOD_ALLI_HATE          = 68, // Range around target to add alliance member to enmity list.
     MOBMOD_NO_LINK            = 69, // If set, mob cannot link until unset.
     MOBMOD_NO_REST            = 70, // Mob cannot regain hp (e.g. re-burrowing antlions during ENM).
+    MOBMOD_LEADER             = 71, // Indicate this mob is a leader. Used for Ul'xzomit mobs.
```
`Used for mobs that follow a defined "leader"` would be preferred description.

TeoTwawki commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -12884,7 +12888,7 @@ uint8 CLuaBaseEntity::getModelSize()
 void CLuaBaseEntity::setMobFlags(uint32 flags, uint32 mobid)
 {
-    XI_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC);
+    XI_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_MOB);
```
m_PBaseEntity here is the player, we get their cursor target in this binding and the mob is passed in from that or an optional ID input.

claywar replied:
Just following up on this as well, we really should move away from these breaks in Lua Land and move towards logging output in debug.

DiscipleOfEris replied:
Ah, that got left in there when I was trying to figure out why it wasn't working the way I was expected at the time. I should have removed the break line entirely, since I need this function for the mob to change its own flag on spawn. Instead, it should just be a check in the function rather than a break.

TeoTwawki replied:
lil trivia:

I had actually been intending for setMobFlags and player nameflag functions and npc's entity flags related functions to all be combined (its the same stuff on diff entities, so its a lot of duplication!) but each entity type is slightly off from the others in where the bits land. so they either need bit shifting or more preferable, rework the masks. There are more bits that belong in those masks that we aren't using - far to much for one single mask field (retail is using EIGHT mask fields for all this crap!!) so its on the back burner till thats done.

TeoTwawki commented on src/map/entities/mobentity.h:
```diff
@@ -57,7 +57,7 @@ enum ROAMFLAG : uint16
     ROAMFLAG_NONE3   = 0x08,  //
     ROAMFLAG_NONE4   = 0x10,  //
-    ROAMFLAG_NONE5   = 0x20,  //
+    ROAMFLAG_SCRIPT  = 0x20,  // Will be implemented in a lua script (e.g. entity.onMobRoamAction())
```
> ROAMFLAG_EVENT   = 0x100, // calls lua method for roaming logic

:thinking: whatever "EVENT" was intended to do its poorly named and described and sounds suspiciously like it was meant to be what you are adding.

Also antlion ambush roamflag value and worm values are suspiciously present in another flag field in retail captures.

Frankly our entire enum is suspect, and I am against adding to it just to do things in script.

----

How pathing normally works is:

Do things in a script say to do things? then do them, else fallback to standard mob behavior. Turning off the script side is itself a scriptable action as well.  If the purpose was to make it toggle-able there shouldn't be a need for more core flags to do that - we have scripts that already do it.

```lua

entity.onPath = function(mob)
    --[[
        path logic here i.e. tell it where it needs to go
        as long as the mob is pathing this will continue to exec each "step" along its path
        example:
            xi.path.patrol(mob, pathNodes, xi.path.flag.RUN)
        tells the pathing logic to grab its points from a table named pathNodes and have "mob" do this 
        with the running flag to "hustle". 

        Hypothetical: Where was momma last tick? add that to our table for next tick,
        and remove where we were last tick.
    ]]
end

entity.onMobRoam = function(mob)
    -- move to start position if not moving
    if not mob:isFollowingPath() then
        --[[
        kick off pathing logic here, in the 1st tick of its roaming.
        once pathing begins standard roam logic shuts off until pathing id manually ended
        or the mobs battle AI begins. Example:
            mob:pathThrough(table input of places to go, any flags needed)
        ]]
    end
end
```

DiscipleOfEris replied:
> ```lua
> entity.onMobRoam = function(mob)
> ```

I didn't use this one because it's hard coded to run once every three seconds. The `onPath()` one can never avoid the code that forces a mob to roam home, which would make `isFollowingPath()` true while roaming home.

TeoTwawki replied:
> I didn't use this one because it's hard coded to run once every three seconds. The `onPath()` one can never avoid the code that forces a mob to roam home, which would make `isFollowingPath()` true while roaming home.

If thats all then we can use a different condition to not re-kick off that function till after it is within a range from home again. For example, a `mob:getLocalVar()` can be read and only kick off pathing if zero, and it within 5 yalms of its homepoint, set it back to zero again. in onPath could set it to one because one in there its pathing.

We could also store the last position it had before it was in battle mode, and path functions do include having it go back to its last node - we just make that pos its last node. Those I do not know how well that works - we haven't been using it except in big tabled NMs.

----

edit:

per discord discussion, a new mobmod for `DONT_ROAM_HOME` is acceptable.

DiscipleOfEris replied:
As we discussed in DMs, I'll make a MobMod that disables the behavior that forces a mob to roam home when it gets a set distance from its spawn location. Then I could refactor the mobs currently using ROAMFLAG_EVENT (dynamis and battlefield mobs) so they use lua to face their spawn location after roaming home (e.g. `onMobRoam() to check distance from spawn and reface its spawn facing when it's 0). Then claim ROAMFLAG_EVENT for my purposes.

I added the `triggerListener()` stuff because I wanted `xi.follow.follow()` to be a fire and forget function, so it could be used in, say, a GM command for example.

DiscipleOfEris replied:
In addition to `DONT_ROAM_HOME` I created one more MobMod: `ROAM_RESET_FACING` which automatically gets set for all the dynamis and battlefield mobs that were relying on `ROAMFLAG_EVENT` to do it before.

DiscipleOfEris committed:
Minor fixes and formatting

Fix an unnecessary error sent to console when an Ul'xzomit leader respawns. Comply with automatic sanity checks.

DiscipleOfEris committed:
Revert `ROAMFLAG_SCRIPT`

Switch to using `ROAMFLAG_EVENT` for this behavior, and add some MobMods to support the old assumptions made with the flag.

DiscipleOfEris committed:
Rename the roamFlag `EVENT` to `SCRIPTED`

The new label is more intuitive and descriptive.

DiscipleOfEris committed:
Comply with Clang styling

Comply with Clang styling

DiscipleOfEris committed:
Improve lua `setMobFlags()`

Allow new usage of `setMobFlags()` so it can be used from a mob, like `mob:setMobFlags(flag)`.

DiscipleOfEris committed:
Implement mobMod `ROAM_RESET_FACING`

For backwards compatibility, this has been added to all mobs that were previously coded to do this. There are currently 147 mobs in the database with the roamFlag `SCRIPTED` set, in addition to battlefield mobs and mobType `EVENT` mobs.

DiscipleOfEris commented:
Damn you, Clang. I'll get you one day.

DiscipleOfEris committed:
Clean redundant function call

DiscipleOfEris commented:
I've updated the text of the PR to have a more up-to-date and thorough summary of the changes.

DiscipleOfEris committed:
Make clusters drops from Ul'xzomit slotted

This should be more compliant with retail.

TeoTwawki commented:
> I've updated the text of the PR to have a more up-to-date and thorough summary of the changes.
> 
> EDIT: I would like feedback on whether we believe the elemental clusters should be slotted drops. The wiki says 10% for clusters. My guess is that means 10% for a kill to result in a cluster (which could be any type), not 10% for each cluster, which means the cluster should be a slotted drop. At the moment, it's 10% per cluster, which means about a 57% chance for at least one cluster to drop.

correct, random clusters sharing a slot on retail. do not rtust wiki drop rates - they combine things that shouldn't be and its all user reported BS. FFXIDB isn't perfect but its at least a little btter.

TeoTwawki commented on sql/mob_spawn_mods.sql:
```diff
@@ -609,6 +609,40 @@ INSERT INTO `mob_spawn_mods` VALUES (17588703,55,180,1);
+-- Ul'xzomit mother roam mods
+INSERT INTO `mob_spawn_mods` VALUES (16912394,31,25,1);
+INSERT INTO `mob_spawn_mods` VALUES (16912409,31,25,1);
+INSERT INTO `mob_spawn_mods` VALUES (16912422,31,25,1);
```
~~could you add `-- mobmod: modnamehere` to at least the 1st row on these 2 groups of mods, so future readers don't have to jump over to the header file?~~ nvm, automation to the rescue

TeoTwawki replied:
other than that, I think you are set here if nobody spots something I missed

claywar replied:
Apologies, but going to pretty much guarantee a merge conflict here after adding standardized comments.

TeoTwawki commented on scripts/zones/AlTaieu/mobs/Ulxzomit.lua:
```diff
@@ -0,0 +1,71 @@
+end
+
+-- Do the following and linking for the babies.
+entity.onMobRoamAction = function(mob)
```
see #2629 and comment there, because any change done for that issue will affect the future usage of roaming scripts and I don't want to break your work.
